docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,387 @@
+
+# Table of Contents
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Guide-level explanation](#guide-level-explanation)
+    - [Conversion classes](#conversion-classes)
+        - [Trivial](#trivial)
+        - [Validated](#validated)
+        - [Cast](#cast)
+        - [Assisted](#assisted)
+    - [Two-phase change](#two-phase-change)
+    - [Key constraints](#key-constraints)
+        - [Interleaved tables](#interleaved-tables)
+    - [Indexed columns](#indexed-columns)
+    - [Views](#views)
+- [Reference-level explanation](#reference-level-explanation)
+    - [Process](#process)
+    - [Failure modes](#failure-modes)
+    - [Job control](#job-control)
+    - [Allowable conversions](#allowable-conversions)
+        - [Array conversions](#array-conversions)
+    - [Drawbacks](#drawbacks)
+    - [Rationale and Alternatives](#rationale-and-alternatives)
+
+
+# Summary
+
+Users should be able to execute DDL statements to alter the type of
+existing columns, without resorting to a copy-and-rename approach. These
+kinds of statements are often used by ORM or other SQL tools or by users
+during their prototyping phase.
+
+This RFC describes a staged approach to enabling column-type changes by
+grouping type conversions by implementation complexity.
+
+The implementation described here supports keeping the table online,
+with some reasonable restrictions applied while the schema change is in
+progress. Changes to primary-key columns are explicitly out of scope for
+this RFC.
+
+# Motivation
+
+Schema evolution is an important part of any database with a non-trivial
+lifetime. While it is possible to effectively change a column's type
+using a copy-and-rename approach, users expect to be able to perform
+in-place changes to a column's type. Users should also have an easy way
+to take advantage of new features in CockroachDB, such as the `INET` or
+`JSONB` types.
+
+# Guide-level explanation
+
+The SQL syntax for this feature is borrowed from [PostgreSQL's `ALTER
+TABLE` syntax](https://www.postgresql.org/docs/9.3/static/sql-altertable.html).
+
+```
+ALTER TABLE table_name
+  ALTER [COLUMN] column_name
+  [SET DATA] TYPE data_type
+  [COLLATE collation]
+  [USING expression]
+```
+
+In general, we assume that a column-type change will not be an
+instantaneous process and could involve a non-trivial amount of wall
+time in order to complete. Column-type changes build on top of existing
+infrastructure to allow a column-type change to occur without taking the
+affected table/column offline. While a column-type change is in
+progress, reads from that column will return the original type, while
+any writes must be compatible with both the current and future schema.
+Once the type-change work has been completed, the new type becomes
+effective and the old data can be retired. We allow only a single
+column-type change to be in progress for any given table at a time;
+multiple type changes will be queued and executed sequentially.
```

Comment:
you can add, "although this design doesn't preclude simultaneous execution of multiple type changes"